ironfish/src/indexer/minedBlocksIndexer.ts

```diff
@@ -0,0 +1,245 @@
+import { Indexer } from './indexer'
+
+type MinedBlocksDBMeta = {
+  headHash: string | null
+}
+export class MinedBlocksIndexer extends Indexer {
+  protected meta: IDatabaseStore<{
+    key: keyof MinedBlocksDBMeta
+    value: MinedBlocksDBMeta[keyof MinedBlocksDBMeta]
+  }>
+  protected minedBlocks: IDatabaseStore<{
+    key: string
+    value: { main: boolean; hash: string; sequence: number; account: string; minersFee: number }
+  }>
+
+  protected readonly accounts: Accounts
+  readonly chain: Blockchain
+  protected readonly chainProcessor: ChainProcessor
+  protected readonly logger: Logger
+  protected isOpen: boolean
+  protected isStarted: boolean
+  protected eventLoopTimeout: SetTimeoutToken | null = null
+
+  constructor({
+    files,
+    location,
+    accounts,
+    chain,
+    logger = createRootLogger(),
+  }: {
+    files: FileSystem
+    location: string
+    accounts: Accounts
+    chain: Blockchain
+    logger?: Logger
+  }) {
+    super({ files, location })
+    this.accounts = accounts
+    this.chain = chain
+    this.logger = logger
+    this.chainProcessor = new ChainProcessor({ logger, chain, head: null })
+    this.isOpen = false
+    this.isStarted = false
+
+    this.meta = this.database.addStore<{
+      key: keyof MinedBlocksDBMeta
+      value: MinedBlocksDBMeta[keyof MinedBlocksDBMeta]
+    }>({
+      name: 'meta',
+      keyEncoding: new StringEncoding<keyof MinedBlocksDBMeta>(),
+      valueEncoding: new JsonEncoding(),
+    })
+
+    this.minedBlocks = this.database.addStore<{
+      key: string
+      value: {
+        main: boolean
+        hash: string
+        sequence: number
+        account: string
+        minersFee: number
+      }
+    }>({
+      name: 'minedBlocks',
+      keyEncoding: new StringEncoding(),
+      valueEncoding: new JsonEncoding(),
+    })
+
+    this.chainProcessor.onAdd.on(async (header) => {
+      const block = await chain.getBlock(header)
+      Assert.isNotNull(block)
+
+      const account = this.accounts.listAccounts().find((a) => isBlockMine(block, a))
+      if (account) {
+        const main = await this.chain.isHeadChain(header)
+        const minersFee = this.chain.strategy.miningReward(header.sequence)
+
+        await this.put(header, account.name, main, minersFee)
+      }
+
+      await this.updateHeadHash(header.hash)
+    })
+
+    this.chainProcessor.onRemove.on(async (header) => {
+      if (await this.minedBlocks.has(header.hash.toString('hex'))) {
+        const block = await chain.getBlock(header)
+        Assert.isNotNull(block)
+
+        const account = this.accounts.listAccounts().find((a) => isBlockMine(block, a))
+        if (account) {
+          const main = await this.chain.isHeadChain(header)
+          const minersFee = this.chain.strategy.miningReward(header.sequence)
+
+          await this.put(header, account.name, main, minersFee)
+        }
+      }
+      await this.updateHeadHash(header.previousBlockHash)
```

Comment:
what's the implication of not doing all this inside of a transaction?

Reply:
Changes would not be atomic, which could lead to non-linearizability. Adding the transaction.

Reply:
I'm unresolving, because a transaction wasn't added.